### PR TITLE
Tiger merge 2022-02-10

### PR DESCRIPTION
**Pull request by kevinkim9264**
Notable additions:
- Added parallel capabilities on notebook exports and imports
- MLflow experiment export/import is supported (Other ML objects will be supported in later batch)

How is this tested?:
```
python3 migration_pipeline.py --profile mwc_src --export-pipeline --use-checkpoint --num-parallel 20
.....
.....
Export finished successfully. Session Id: M20220210152603
2022-02-10,15:40:42;INFO;finish_export Completed. Total time taken: 0:00:00.016202

```

```
python3 migration_pipeline.py --profile mwc_dst --import-pipeline --use-checkpoint --num-parallel 20 --session M20220210152603
......
.....
Successfull notebook run with notebook_output: {"total_num_acls": 15, "num_sucessfully_executed": 15, "num_execution_errors": 0, "num_error_entries_acls": 0}
2022-02-10,16:05:20;INFO;import_metastore_table_acls Completed. Total time taken: 0:05:28.522699


```

### CODE REVIEW

File: test/thread_safe_writer_test.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+        fp1 = open(f1, "r")
+        fp2 = open(f2, "r")
+        f1_lines = fp1.readlines()
+        f2_lines = fp2.readlines()
+        fp1.close()
+        fp2.close()
+
+        # since it is multi thread writing to the same file, the order is not guaranteed.
+        # hence we test the content equality by sorting and then comparing.
+        assert(f1_lines.sort() == f2_lines.sort())
```
Comment (paulg-mscience) [inline test/thread_safe_writer_test.py:49]: Nice

**Comment by paulg-mscience** [comment]
@kevinkim9264 Nice work on this pull request! I am testing it against our Databricks workspace. Exporting with num-parallel = 30 looks great except for `export_workspace_items_log`. I have studied your code but cannot see why this task has gone from 7 minutes taking 6+ hours.

Here are my timings:
| Task| master branch | pull request branch     |
| :---        |   ---:   |          ---: |
|export_instance_profiles|0:00:00|0:00:00|
|export_users|0:00:18|0:00:19|
|export_groups|0:00:26|0:00:45|
|**export_workspace_items_log**|0:07:30|**6:39:00**|
|Notebook ACLs Export|1:18:57|0:32:45|
|Directories ACLs Export|0:13:53|0:05:15|
|export_notebooks|3:54:20|0:39:32|
|export_secrets|0:05:04|0:06:43|
|export_clusters|0:00:03|0:00:08|
|export_instance_pools|0:00:00|0:00:00|
|export_jobs|0:00:53|0:01:59|
|export_metastore|skipped|skipped|
|export_metastore_table_acls|skipped|skipped|
|finish_export|0:00:00|0:00:00|

**Comment by kevinkim9264** [comment]
Thank you for the testing and the results!

I am very surprised by the result from the workspace-item exporting time. That is the step that should benefit the most from the parallelization, because it is not even bottlenecked by single file writing. 

Could you tell me which command you ran to export the workspace-item? Also could you run that command on num-parallel=20?

Best,
Kevin

**Comment by kevinkim9264** [comment]
Also did you get a chance do some testing for import parts?

**Comment by paulg-mscience** [comment]
> Also did you get a chance do some testing for import parts?
>
Yes, I am testing the import as well. I have yet to get a successful import, but as soon as I do I will report my findings.

**Comment by kevinkim9264** [comment]
@paulg-mscience could you try running
```
python3 export_db.py --profile old --workspace  --num-parallel 30
```
Unfortunately I couldn't reproduce the regression on the exporting.

**Comment by paulg-mscience** [comment]
@kevinkim9264 
Ran this command from `master` branch and it wrote 90,7474 lines to the user_workspace.log in 12 mins.
```
date; time python export_db.py --profile old --workspace
```
Ran this command from `kevinkim9264:tiger-merge-2022-02-10` branch and it wrote 31,311 lines to user_workspace.log in 55 mins before I aborted.
```
date; time python export_db.py --profile old --workspace  --use-checkpoint --num-parallel 30
```

**Comment by kevinkim9264** [comment]
@paulg-mscience thanks for the response. I am wondering if it is due to the high number of threads contending the same file, which I can't reproduce wierdly.

Could you do me a favor and try different number of num-parallel and give the time? (feel free to abort after too much time)
```
date; time python export_db.py --profile old --workspace  --use-checkpoint --num-parallel 1

date; time python export_db.py --profile old --workspace  --use-checkpoint --num-parallel 5

date; time python export_db.py --profile old --workspace  --use-checkpoint --num-parallel 10

date; time python export_db.py --profile old --workspace  --use-checkpoint --num-parallel 20
```

**Comment by paulg-mscience** [comment]
@kevinkim9264 
I agree that this looks like threading contention of some kind. Please be aware that our notebook folder structure is heavily nested which I believe is a requirement to reproduce the slowness.

I ran two tests which I aborted because they were progressing slowly.
```
date; time python export_db.py --profile old --workspace  --use-checkpoint --num-parallel 1
17m8.887s
3048 records written to user_workspace.log
```
```
date; time python export_db.py --profile old --workspace  --use-checkpoint --num-parallel 5
13m47.520s
8471 records written to user_workspace.log
```
When I aborted this second test the program dumped call stacks for all threads. Below is a sampling. There were too many to call stacks to paste here.
```
Thread 0x00007fa863e9f700 (most recent call first):
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/logging/__init__.py", line 899 in acquire
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/logging/__init__.py", line 948 in handle
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/logging/__init__.py", line 1657 in callHandlers
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/logging/__init__.py", line 1595 in handle
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/logging/__init__.py", line 1585 in _log
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/logging/__init__.py", line 1442 in info
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/logging/__init__.py", line 2078 in info
  File "/home/ec2-user/github/databricks-migrate/dbclient/WorkspaceClient.py", line 425 in log_all_workspace_items
  File "/home/ec2-user/github/databricks-migrate/dbclient/WorkspaceClient.py", line 436 in _recurse_log_all_workspace_items
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/concurrent/futures/thread.py", line 57 in run
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/concurrent/futures/thread.py", line 80 in _worker
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/threading.py", line 870 in run
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/threading.py", line 932 in _bootstrap_inner
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/threading.py", line 890 in _bootstrap

Thread 0x00007fa7d37b8700 (most recent call first):
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/logging/__init__.py", line 899 in acquire
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/logging/__init__.py", line 948 in handle
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/logging/__init__.py", line 1657 in callHandlers
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/logging/__init__.py", line 1595 in handle
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/logging/__init__.py", line 1585 in _log
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/logging/__init__.py", line 1442 in info
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/logging/__init__.py", line 2078 in info
  File "/home/ec2-user/github/databricks-migrate/dbclient/WorkspaceClient.py", line 425 in log_all_workspace_items
  File "/home/ec2-user/github/databricks-migrate/dbclient/WorkspaceClient.py", line 436 in _recurse_log_all_workspace_items
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/concurrent/futures/thread.py", line 57 in run
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/concurrent/futures/thread.py", line 80 in _worker
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/threading.py", line 870 in run
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/threading.py", line 932 in _bootstrap_inner
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/threading.py", line 890 in _bootstrap

Thread 0x00007fa7d41b9700 (most recent call first):
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/threading.py", line 302 in wait
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/threading.py", line 558 in wait
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/concurrent/futures/_base.py", line 244 in as_completed
  File "/home/ec2-user/github/databricks-migrate/dbclient/WorkspaceClient.py", line 444 in log_all_workspace_items
  File "/home/ec2-user/github/databricks-migrate/dbclient/WorkspaceClient.py", line 436 in _recurse_log_all_workspace_items
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/concurrent/futures/thread.py", line 57 in run
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/concurrent/futures/thread.py", line 80 in _worker
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/threading.py", line 870 in run
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/threading.py", line 932 in _bootstrap_inner
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/threading.py", line 890 in _bootstrap

Thread 0x00007fa7eb3de700 (most recent call first):
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/logging/__init__.py", line 899 in acquire
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/logging/__init__.py", line 948 in handle
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/logging/__init__.py", line 1657 in callHandlers
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/logging/__init__.py", line 1595 in handle
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/logging/__init__.py", line 1585 in _log
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/logging/__init__.py", line 1442 in info
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/logging/__init__.py", line 2078 in info
  File "/home/ec2-user/github/databricks-migrate/dbclient/WorkspaceClient.py", line 425 in log_all_workspace_items
  File "/home/ec2-user/github/databricks-migrate/dbclient/WorkspaceClient.py", line 436 in _recurse_log_all_workspace_items
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/concurrent/futures/thread.py", line 57 in run
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/concurrent/futures/thread.py", line 80 in _worker
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/threading.py", line 870 in run
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/threading.py", line 932 in _bootstrap_inner
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/threading.py", line 890 in _bootstrap

Thread 0x00007fa8eeb7d700 (most recent call first):
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/threading.py", line 302 in wait
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/threading.py", line 558 in wait
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/concurrent/futures/_base.py", line 244 in as_completed
  File "/home/ec2-user/github/databricks-migrate/dbclient/WorkspaceClient.py", line 444 in log_all_workspace_items
  File "/home/ec2-user/github/databricks-migrate/dbclient/WorkspaceClient.py", line 436 in _recurse_log_all_workspace_items
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/concurrent/futures/thread.py", line 57 in run
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/concurrent/futures/thread.py", line 80 in _worker
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/threading.py", line 870 in run
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/threading.py", line 932 in _bootstrap_inner
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/threading.py", line 890 in _bootstrap

Thread 0x00007fa81d42e700 (most recent call first):
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/threading.py", line 302 in wait
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/threading.py", line 558 in wait
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/concurrent/futures/_base.py", line 244 in as_completed
  File "/home/ec2-user/github/databricks-migrate/dbclient/WorkspaceClient.py", line 444 in log_all_workspace_items
  File "/home/ec2-user/github/databricks-migrate/dbclient/WorkspaceClient.py", line 436 in _recurse_log_all_workspace_items
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/concurrent/futures/thread.py", line 57 in run
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/concurrent/futures/thread.py", line 80 in _worker
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/threading.py", line 870 in run
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/threading.py", line 932 in _bootstrap_inner
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/threading.py", line 890 in _bootstrap

Thread 0x00007fa856289700 (most recent call first):
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/threading.py", line 1027 in _wait_for_tstate_lock
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/threading.py", line 1011 in join
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/concurrent/futures/thread.py", line 236 in shutdown
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/concurrent/futures/_base.py", line 636 in __exit__
  File "/home/ec2-user/github/databricks-migrate/dbclient/WorkspaceClient.py", line 447 in log_all_workspace_items
  File "/home/ec2-user/github/databricks-migrate/dbclient/WorkspaceClient.py", line 436 in _recurse_log_all_workspace_items
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/concurrent/futures/thread.py", line 57 in run
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/concurrent/futures/thread.py", line 80 in _worker
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/threading.py", line 870 in run
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/threading.py", line 932 in _bootstrap_inner
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/threading.py", line 890 in _bootstrap

Thread 0x00007fa7d73be700 (most recent call first):
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/logging/__init__.py", line 899 in acquire
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/logging/__init__.py", line 948 in handle
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/logging/__init__.py", line 1657 in callHandlers
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/logging/__init__.py", line 1595 in handle
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/logging/__init__.py", line 1585 in _log
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/logging/__init__.py", line 1442 in info
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/logging/__init__.py", line 2078 in info
  File "/home/ec2-user/github/databricks-migrate/dbclient/WorkspaceClient.py", line 425 in log_all_workspace_items
  File "/home/ec2-user/github/databricks-migrate/dbclient/WorkspaceClient.py", line 436 in _recurse_log_all_workspace_items
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/concurrent/futures/thread.py", line 57 in run
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/concurrent/futures/thread.py", line 80 in _worker
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/threading.py", line 870 in run
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/threading.py", line 932 in _bootstrap_inner
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/threading.py", line 890 in _bootstrap

Thread 0x00007fa8a1d02700 (most recent call first):
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/threading.py", line 302 in wait
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/threading.py", line 558 in wait
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/concurrent/futures/_base.py", line 244 in as_completed
  File "/home/ec2-user/github/databricks-migrate/dbclient/WorkspaceClient.py", line 444 in log_all_workspace_items
  File "/home/ec2-user/github/databricks-migrate/dbclient/WorkspaceClient.py", line 436 in _recurse_log_all_workspace_items
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/concurrent/futures/thread.py", line 57 in run
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/concurrent/futures/thread.py", line 80 in _worker
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/threading.py", line 870 in run
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/threading.py", line 932 in _bootstrap_inner
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/threading.py", line 890 in _bootstrap

Thread 0x00007fa86caad700 (most recent call first):
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/logging/__init__.py", line 899 in acquire
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/logging/__init__.py", line 948 in handle
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/logging/__init__.py", line 1657 in callHandlers
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/logging/__init__.py", line 1595 in handle
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/logging/__init__.py", line 1585 in _log
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/logging/__init__.py", line 1442 in info
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/logging/__init__.py", line 2078 in info
  File "/home/ec2-user/github/databricks-migrate/dbclient/WorkspaceClient.py", line 425 in log_all_workspace_items
  File "/home/ec2-user/github/databricks-migrate/dbclient/WorkspaceClient.py", line 436 in _recurse_log_all_workspace_items
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/concurrent/futures/thread.py", line 57 in run
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/concurrent/futures/thread.py", line 80 in _worker
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/threading.py", line 870 in run
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/threading.py", line 932 in _bootstrap_inner
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/threading.py", line 890 in _bootstrap

Thread 0x00007fa86d4ae700 (most recent call first):
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/logging/__init__.py", line 899 in acquire
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/logging/__init__.py", line 948 in handle
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/logging/__init__.py", line 1657 in callHandlers
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/logging/__init__.py", line 1595 in handle
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/logging/__init__.py", line 1585 in _log
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/logging/__init__.py", line 1442 in info
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/logging/__init__.py", line 2078 in info
  File "/home/ec2-user/github/databricks-migrate/dbclient/WorkspaceClient.py", line 425 in log_all_workspace_items
  File "/home/ec2-user/github/databricks-migrate/dbclient/WorkspaceClient.py", line 436 in _recurse_log_all_workspace_items
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/concurrent/futures/thread.py", line 57 in run
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/concurrent/futures/thread.py", line 80 in _worker
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/threading.py", line 870 in run
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/threading.py", line 932 in _bootstrap_inner
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/threading.py", line 890 in _bootstrap

Thread 0x00007fa8df164700 (most recent call first):
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/logging/__init__.py", line 899 in acquire
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/logging/__init__.py", line 948 in handle
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/logging/__init__.py", line 1657 in callHandlers
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/logging/__init__.py", line 1595 in handle
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/logging/__init__.py", line 1585 in _log
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/logging/__init__.py", line 1442 in info
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/logging/__init__.py", line 2078 in info
  File "/home/ec2-user/github/databricks-migrate/dbclient/WorkspaceClient.py", line 415 in log_all_workspace_items
  File "/home/ec2-user/github/databricks-migrate/dbclient/WorkspaceClient.py", line 436 in _recurse_log_all_workspace_items
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/concurrent/futures/thread.py", line 57 in run
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/concurrent/futures/thread.py", line 80 in _worker
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/threading.py", line 870 in run
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/threading.py", line 932 in _bootstrap_inner
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/threading.py", line 890 in _bootstrap

Thread 0x00007fa8e0f67700 (most recent call first):
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/logging/__init__.py", line 1065 in flush
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/logging/__init__.py", line 1085 in emit
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/logging/__init__.py", line 950 in handle
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/logging/__init__.py", line 1657 in callHandlers
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/logging/__init__.py", line 1595 in handle
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/logging/__init__.py", line 1585 in _log
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/logging/__init__.py", line 1442 in info
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/logging/__init__.py", line 2078 in info
  File "/home/ec2-user/github/databricks-migrate/dbclient/WorkspaceClient.py", line 415 in log_all_workspace_items
  File "/home/ec2-user/github/databricks-migrate/dbclient/WorkspaceClient.py", line 436 in _recurse_log_all_workspace_items
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/concurrent/futures/thread.py", line 57 in run
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/concurrent/futures/thread.py", line 80 in _worker
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/threading.py", line 870 in run
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/threading.py", line 932 in _bootstrap_inner
  File "/home/ec2-user/.pyenv/versions/3.8.7/lib/python3.8/threading.py", line 890 in _bootstrap
```

**Comment by kevinkim9264** [comment]
@paulg-mscience could you run it again and see if the regression is gone? I removed the concurrency logic for --workspace because it wasn't showing much improvements in ideal case anyway as explained here: https://github.com/databrickslabs/migrate/pull/103 (due to the bottleneck on the fact that only a single thread can write to the file). 

In your case, because of the recursive logic, it seems like it was spawning more threads in faster rate than the rate it closes the thread (and makes sense in the heavily nested directory structure case)

**Comment by kevinkim9264** [comment]
@paulg-mscience thank you! I incorporated your change into this PR and did some testing. please take another look and let me know if anything is not working

**Review by mrchristine — APPROVED** [review]
Overall looks good to me. Tested locally on my workspaces and functionally it works.

File: dbclient/parser.py [hunk 2]
```diff
@@ -295,10 +302,14 @@ def get_import_parser():
...
                         help='Skip missing users that do not exist when importing user notebooks')
 
-    # get all secret scopes
+    # import all secret scopes
     parser.add_argument('--secrets', action='store_true',
                         help='Import all secret scopes')
 
+    # import all mlflow experiments
+    parser.add_argument('--mlflow-experiments', action='store_true',
+                        help='log all the mlflow experiments')
```
[the remark this answers is not in the document]
Comment (kevinkim9264) [reply dbclient/parser.py:311]: will do on the later push. thanks

**Pull request merged**